# Data Visualization (2018/19)

## Solutions for Assignment 2 - Exploratory Data Analysis

Presented by Group: 53
- Lucia Eve Berger
- Kumail Raza

Date: 22.11.2018, 22:00

Hints: 
- Double click on a cell to edit it.
- Enter the Markdown mode in the drop-down menu above (it may say Code for new cells).
- Run a cell by pressing CTRL+Enter
- Markdown docu: http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html
- Markdown cheatsheet: http://datascience.ibm.com/blog/markdown-for-jupyter-notebooks-cheatsheet/"""

### Exercise 1: Choose a data set

We will use the Python Data Analysis Library (Pandas) to work with the data. Pandas provides a rich feature set for working with tabular data including data structures and analysis tools.

A 10 minute introduction to pandas with all neccessary functions can be found here: http://pandas.pydata.org/pandas-docs/stable/10min.html (Quickly skim over the available functionality. You don't have to learn it yet.)

In [2]:
# import the pandas library and call it pd for further usage
import pandas as pd

#### Sample code to load the baseball data and get basic statistics

In [3]:
# read the csv file into a pandas dataframe and print the first lines of the table
filename = "baseball_data.csv"
df = pd.read_csv( filename, header=0 )
df.head()

name handedness  height  weight    avg   HR
0       Tom Brown          R      73     170  0.000    0
1  Denny Lemaster          R      73     182  0.130    4
2       Joe Nolan          L      71     175  0.263   27
3     Denny Doyle          L      69     175  0.250   16
4   Jose Cardenal          R      70     150  0.275  138

In [4]:
df.describe(include="all")

name handedness       height       weight          avg  \
count             1157       1157  1157.000000  1157.000000  1157.000000   
unique            1151          3          NaN          NaN          NaN   
top     Dave Stapleton          R          NaN          NaN          NaN   
freq                 2        737          NaN          NaN          NaN   
mean               NaN        NaN    72.756266   184.513397     0.186793   
std                NaN        NaN     2.142272    15.445995     0.106175   
min                NaN        NaN    65.000000   140.000000     0.000000   
25%                NaN        NaN    71.000000   175.000000     0.138000   
50%                NaN        NaN    73.000000   185.000000     0.238000   
75%                NaN        NaN    74.000000   195.000000     0.258000   
max                NaN        NaN    80.000000   245.000000     0.328000   

                 HR  
count   1157.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean      45.359551  
std       74.065110  
min        0.000000  
25%        1.000000  
50%       15.000000  
75%       55.000000  
max      563.000000

#### Sample code to load the Titanic data and get basic statistics

In [5]:
# read the csv file into a pandas dataframe and print the first lines of the table
filename = "titanic3.csv"
df = pd.read_csv( filename, header=0 )
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [6]:
df.dtypes

pclass         int64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object

In [7]:
df.describe(include="all")

pclass     survived              name   sex          age  \
count   1309.000000  1309.000000              1309  1309  1046.000000   
unique          NaN          NaN              1307     2          NaN   
top             NaN          NaN  Kelly, Mr. James  male          NaN   
freq            NaN          NaN                 2   843          NaN   
mean       2.294882     0.381971               NaN   NaN    29.881138   
std        0.837836     0.486055               NaN   NaN    14.413493   
min        1.000000     0.000000               NaN   NaN     0.170000   
25%        2.000000     0.000000               NaN   NaN    21.000000   
50%        3.000000     0.000000               NaN   NaN    28.000000   
75%        3.000000     1.000000               NaN   NaN    39.000000   
max        3.000000     1.000000               NaN   NaN    80.000000   

              sibsp        parch    ticket         fare        cabin embarked  \
count   1309.000000  1309.000000      1309  1308.000000          295     1307   
unique          NaN          NaN       929          NaN          186        3   
top             NaN          NaN  CA. 2343          NaN  C23 C25 C27        S   
freq            NaN          NaN        11          NaN            6      914   
mean       0.498854     0.385027       NaN    33.295479          NaN      NaN   
std        1.041658     0.865560       NaN    51.758668          NaN      NaN   
min        0.000000     0.000000       NaN     0.000000          NaN      NaN   
25%        0.000000     0.000000       NaN     7.895800          NaN      NaN   
50%        0.000000     0.000000       NaN    14.454200          NaN      NaN   
75%        1.000000     0.000000       NaN    31.275000          NaN      NaN   
max        8.000000     9.000000       NaN   512.329200          NaN      NaN   

       boat        body     home.dest  
count   486  121.000000           745  
unique   27         NaN           369  
top      13         NaN  New York, NY  
freq     39         NaN            64  
mean    NaN  160.809917           NaN  
std     NaN   97.696922           NaN  
min     NaN    1.000000           NaN  
25%     NaN   72.000000           NaN  
50%     NaN  155.000000           NaN  
75%     NaN  256.000000           NaN  
max     NaN  328.000000           NaN

#### Your selection:
Selected dataset: Titanic

### Exercise 2: Data queries

1. Where were the passengers travelling to? What are their home destinations?
2. Did class influence survival rate of passengers? Did Class 1 passengers survive more?
3. Did age and/or sex influence the rate of survival of passengers?
4. Influence of family size on survival rate.
5. Age distribution within class?
6. How did the location of the passengers (cabin no) influence their survival rate?

### Exercise 3: Data analysis

Here is sample code to pass your data from the pandas data frame to bokeh and render a graphic inside the notebook.

In [8]:
filename = "titanic3.csv"
df = pd.read_csv( filename, header=0 )

In [9]:
from math import pi
import bokeh 
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource
from bokeh.layouts import row, column
# import the colours for the graphs 
from bokeh.palettes import Blues, Greys
from bokeh.transform import cumsum
# allow additional graphing features (Box Annotation)
from bokeh.core.properties import value
from bokeh.models import ColumnDataSource,BoxAnnotation
from bokeh.transform import dodge
from bokeh.io import output_notebook

output_notebook()

Loading BokehJS ...

In [10]:
def create_bar_chart():
    df_home_destinations = (df['home.dest'].value_counts().to_frame('count'))[:25]
    df_home_destinations.head()
    list_of_places = list(df_home_destinations.index)
    p = figure(x_range=list_of_places, width=1000,title="Top 25 Geographical Destinations of Passengers" )
    p.vbar(source=df_home_destinations, x='index', width=1.0, bottom=0, top='count',
            line_color='white', fill_color='cornflowerblue')
    
    # rotates the axis to increase legibilitity 
    p.y_range.start = 0
    p.title.align = 'center'
    p.xaxis.major_label_orientation = "vertical"
    p.x_range.range_padding = 0.1
    p.xgrid.grid_line_color = None
    p.axis.minor_tick_line_color = None
    p.outline_line_color = None
    p.xaxis.axis_label = "Desination [name]"
    p.yaxis.axis_label = "count"
    p.xgrid.visible = False
    show(p)


create_bar_chart()

In [11]:
def create_class_sp():
    rates_of_survival = {'1': 0, '2': 0, '3': 0}
    rates_of_death = {'1': 0, '2': 0, '3': 0}
    for i in range(1, 4):
        rates_of_survival[str(i)] = \
            len(survival_rate.loc[survival_rate['pclass'] == i])/len(survival_rate)
        rates_of_death[str(i)] = \
            len(death_rate.loc[death_rate['pclass'] == i])/len(death_rate)
    death_chart = create_wedge_chart(rates_of_death, Greys, "Rate of Mortality by Class (Total Rate: {} %)".format(total_death_rate))
    survival_chart = create_wedge_chart(rates_of_survival, Blues, "Rate of Survival by Class (Total Rate: {} %)".format(total_survival_rate))
    show(row(survival_chart,death_chart))

def create_wedge_chart(data, colors, title):
    data = pd.Series(data).reset_index(name='value').rename(columns={'index': 'rates_of_death'})
    data["legend"] = "Economic Class: " + (data['rates_of_death'].astype(str)) + " (" + (round(data['value'],3).astype(str)) + "%)"                        
    data['angle'] = data['value'] / data['value'].sum() * 2 * pi
    data['color'] = colors[len(data)]
    p = figure(width=500, title=title, toolbar_location=None,
               tools="hover", tooltips="Class: @rates_of_death: @value %", height=350)
    
    p.wedge(x=0, y=1, radius=0.4,
            start_angle=cumsum('angle', include_zero=True), end_angle=cumsum('angle'),
            line_color="white", fill_color='color', legend='legend', source=data)
    p.title.align = 'center'
    p.axis.axis_label=None
    p.axis.visible=False
    p.grid.grid_line_color = None
    return p


def create_horizonal_chart():
    economic_classes = ['Class 1 ', 'Class 2 ', 'Class 3']
    time_slots = ["1", "2", "3"]
    survival = {'economic_classes' : economic_classes,
               '1': [200, 0, 0],
               '2': [0, 119, 0],
               '3': [0, 0, 181],
                'color': Blues[len(economic_classes)]
                }
    death = {'economic_classes' : economic_classes,
               '1': [-123, 0, 0],
               '2': [0, -158, 0],
               '3': [0, 0, -528],
               'color': Greys[len(economic_classes)]
             } 
    p = figure(y_range=economic_classes, width=1000, height=400,x_range=(-600, 600), 
               title="Count of Survivors by Economic Class (death as negative)")

    p.hbar_stack(time_slots, y='economic_classes', height=0.8, color='color', source=ColumnDataSource(death))
    p.hbar_stack(time_slots, y='economic_classes', height=0.8, color="color", source=ColumnDataSource(survival))
    p.y_range.range_padding = 0.1
    p.ygrid.grid_line_color = None
    p.axis.minor_tick_line_color = None
    p.outline_line_color = None
    p.title.align = 'center'
    p.xaxis.axis_label = "Count of Mortality (Negative), Count of Survival (Positive) "
    p.yaxis.axis_label = "Grouped by Class"
    show(p)

# queries on the population of survived
survival_rate = df.loc[df['survived'] == 1]
total_survival_rate = round(survival_rate['survived'].count()/df['survived'].count(),3)
death_rate = df.loc[df['survived'] == 0]
total_death_rate = round(death_rate['survived'].count()/df['survived'].count(),3)

# calls functions to render the graphs
create_class_sp()
horizonal_chart = create_horizonal_chart()

In [12]:
def survival_rate_by_age():
    age_sorted = df.sort_values(by='age')
    columns = ['age_bracket', 'count', 'survival', 'male', 'female', 'male_percentage', 'female_percentage']
    lst = list()
    for i in range(1, 90, 10):
        subset = age_sorted.loc[(age_sorted["age"] < i + 10) & (age_sorted["age"] > i - 10)]
        count_of_male = len(subset.loc[
            (subset['sex'] == 'male') & (subset['survived'] == 1)])
        male_total = len(subset.loc[(subset['sex'] == 'male')])
        female_total = len(subset.loc[(subset['sex'] == 'female')])
        count_of_female = len(subset.loc[
                                (subset['sex'] == 'female') & (subset['survived'] == 1)])
        lst.append([i, len(subset), len(subset.loc[subset['survived'] == 1]) / len(subset),
                    count_of_male, count_of_female, count_of_male/male_total, count_of_female/female_total])
    
    age_survival_dict = pd.DataFrame(lst, columns=columns)
    return age_survival_dict
    
    
def render_graphs(data_input):
    line_chart = figure(title="Rate of Survival over Age", width=500,height=400)
    line_chart.line(source=data_input, x='age_bracket', y='survival', line_width=2)
    age_cohorts = ['0-10', '11-20', '20-30', '30-40', '40-50', '50-60', '60-70', '70-80', '80+']
    line_chart.xaxis.axis_label = "Age"
    line_chart.yaxis.axis_label = "Survival Rates (%)"
    line_chart.title.align = 'center'
    under_50 = BoxAnnotation(top=0.5, fill_alpha=0.1, fill_color='red')
    line_chart.add_layout(under_50)
    
    data = {'age_cohorts': age_cohorts,
            'male': data_input['male'],
            'female': data_input['female'], 
            'male_percentage': data_input['male_percentage'],
            'female_percentage': data_input['female_percentage']
            }
    
    source = ColumnDataSource(data=data)
    p = figure(width=500, height=400, x_range=age_cohorts, y_range=(0, 1), title="Survival Rate by Sex and Age",
               toolbar_location=None, tools="")
    
    p.vbar(x=dodge('age_cohorts', -0.25, range=p.x_range), top='male_percentage', width=0.2, source=source,
           color="#c9d9d3", legend=value("male survival (%)"))
    
    p.vbar(x=dodge('age_cohorts', 0.0, range=p.x_range), top='female_percentage', width=0.2, source=source,
           color="#718dbf", legend=value("female survival (%)"))
    
    p.y_range.start = 0
    
    p.x_range.range_padding = 0.1
    p.xgrid.grid_line_color = None
    p.xaxis.axis_label = 'Age Cohorts'
    p.yaxis.axis_label = 'Survival Rates (%)'
    p.axis.minor_tick_line_color = None
    p.outline_line_color = None
    p.legend.location = "top_left"
    p.legend.orientation = "horizontal"
    p.title.align = 'center'
    show(row(line_chart, p))
    
    
age_survival_dict = survival_rate_by_age()
age_survival_dict.head()
render_graphs(age_survival_dict)


In [16]:
from bokeh.core.properties import value
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure
from bokeh.transform import dodge
import numpy as np
from bokeh.models import HoverTool

output_notebook()

Loading BokehJS ...

In [17]:
csurv_data = df[df['survived'] == 1].groupby('sibsp')['survived'].count()
cnsurv_data = df[df['survived'] == 0].groupby('sibsp')['survived'].count()
total = csurv_data.sum() + cnsurv_data.sum()

fam_total = []
perc_sur = []
perc_nsur = []

cs_val1 = np.array(csurv_data.get_values())
cs_val2 = np.array(cnsurv_data.get_values())
cs_names = np.array(cnsurv_data.keys())
cs_val1 = cs_val1.tolist()
cs_val1.append(0)
cs_val1.append(0)

for x,y in zip(cs_val1,cs_val2):
    fam_total.append(x+y)

for x,y in zip(cs_val1,fam_total):
    perc_sur.append((x/y)*100)

for x,y in zip(cs_val2,fam_total):
    perc_nsur.append((x/y)*100)


family_size = ['Alone','1 FM','2 FM','3 FM','4 FM','5 FM','8 FM']
years = ['Survived', 'Not Survived']

data = {'fam_size' : family_size,
        'Survived'   : perc_sur,
        'Not Survived'   : perc_nsur
       }

source = ColumnDataSource(data=data)

p = figure(x_range=family_size, plot_height=250, title="Survival Rate w.r.t Family Size",
           toolbar_location=None, tools="")

p.vbar(x=dodge('fam_size', -0.3, range=p.x_range), top="Survived", width=0.3, source=source,
       color="#00ff00", legend=value("Survived"))

p.vbar(x=dodge('fam_size',  0.0,  range=p.x_range), top='Not Survived', width=0.3, source=source,
       color="#ff0000", legend=value("Not Survived"))

p.x_range.range_padding = 0.1
p.xgrid.grid_line_color = None
p.legend.location = "top_left"
p.legend.orientation = "horizontal"
p.xaxis.axis_label = "Family Members (FM)"
p.yaxis.axis_label = '% age Passengers'
hover = HoverTool(tooltips=[('Survival %age', '@Survived')
                                    ],
                          mode='vline')

p.add_tools(hover)
show(p)



The chart above shows that survival rate was better for passengers having 1 or 2 family members on board (about 51% and 45% respectively). For the people having more than 5 family members on board, the survival rate is almost zero. This is a strange result (an outlier) as one would expect at least someone from, e.g a family of 8, should've survived. The general inference from the visualization is that if one had a large family on board then it's unlikely that they would've been allowed (the whole family) to board the escape rafts, hence lower survival percentage.

In [24]:
"""Helps to build the age distribution graph"""
def calcAgeSize(df,rl,rg,pc):
    c = df[df['pclass'] == pc]
    age = c[c['age'] <= rg]['age']
    age = np.array(age)
    age = age[age > rl]
    return age.size


In [26]:
from bokeh.core.properties import value
from bokeh.io import show, output_notebook
from bokeh.plotting import figure
import numpy as np
from bokeh.models import HoverTool

output_notebook()

ages = ['0-10', '11-20', '21-30', '31-40', '41-50', '51-60', '61-70', '71-80', '81-90']
years = ["1 Class", "2 Class", "3 Class"]
colors = ["#c9d9d3", "#718dbf", "#e84d60"]

data = {'Ages' : ages,
        '1 Class'   : [calcAgeSize(df,0,10.9,1),calcAgeSize(df,10,20.9,1),calcAgeSize(df,20,30.9,1),calcAgeSize(df,30,40.9,1),calcAgeSize(df,40,50.9,1),\
                       calcAgeSize(df,50,60.9,1), calcAgeSize(df,60,70.9,1), calcAgeSize(df,70,80.9,1), calcAgeSize(df,80,90.9,1)],
        '2 Class'   : [calcAgeSize(df,0,10.9,2),calcAgeSize(df,10,20.9,2),calcAgeSize(df,20,30.9,2),calcAgeSize(df,30,40.9,2),calcAgeSize(df,40,50.9,2),\
                       calcAgeSize(df,50,60.9,2), calcAgeSize(df,60,70.9,2), calcAgeSize(df,70,80.9,2), calcAgeSize(df,80,90.9,2)],
        '3 Class'   : [calcAgeSize(df,0,10.9,3),calcAgeSize(df,10,20.9,3),calcAgeSize(df,20,30.9,3),calcAgeSize(df,30,40.9,3),calcAgeSize(df,40,50.9,3),\
                       calcAgeSize(df,50,60.9,3), calcAgeSize(df,60,70.9,3), calcAgeSize(df,70,80.9,3), calcAgeSize(df,80,90.9,3)]}

p = figure(x_range=ages, plot_height=250, title="Age Distribution of Passenegers within Classes",
           toolbar_location=None, tools="", )

p.vbar_stack(years, x='Ages', width=0.9, color=colors, source=data,
             legend=[value(x) for x in years])

p.y_range.start = 0
p.x_range.range_padding = 0.1
p.xgrid.grid_line_color = None
p.axis.minor_tick_line_color = None
p.outline_line_color = None
p.legend.location = "top_right"
p.legend.orientation = "horizontal"
p.xaxis.axis_label = "Age Range"
p.yaxis.axis_label = 'No. of Passengers'

show(p)

Loading BokehJS ...

The age distribution of passengers yield some interesting observations. Most of infants, kids and teenagers kids were in the 3rd class. This also tells us that the 3rd class was quite populated. The percentage of young adults and middle aged people in the 1st class is almost the same. There were large number of young adults in the 2nd class. There were only a few elderly, all of them were in the third class.

In [20]:
def checkInst(c_dict, c_name, c_val):
    if c_dict.get(c_name) == None:
        c_dict[c_name] = c_val
    else:
        c_dict[c_name] = c_dict.get(c_name) + c_val
        

In [21]:
from collections import defaultdict

def prepCabinData(df):

    df_home = df['cabin']
    n = np.array(df_home.get_values())
    num = 0
    for x in n:
        if isinstance(x, float):
            num +=1



    df_home_destinations = df['cabin'].value_counts()
    c_val = df_home_destinations.get_values()

    cabins = np.array(df_home_destinations.keys())
    c_name = []
    for val in cabins:
        c_name.append(val[0])

    c_dict = defaultdict(int)

    for nam,v in zip(c_name, c_val):
        checkInst(c_dict,nam,v)    
    c_dict['NaN'] = num
    return c_dict


def prepCabinSurvData(df):
    cs_data = df[df['survived'] == 1].groupby('cabin')['survived'].count()
    cs_val = np.array(cs_data.get_values())
    cs_names = np.array(cs_data.keys())
    cs_name = []

    for val in cs_names:
        cs_name.append(val[0])

    cs_dict = defaultdict(int)

    for cs_n, cs_val in zip(cs_name, cs_val):
        checkInst(cs_dict, cs_n, cs_val)
    return cs_dict

In [27]:
from bokeh.models import HoverTool
import numpy as np
output_notebook()

ages = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'T']
years = ["Survived", "Total"]
colors = ["#c9d9d3", "#e84d60"]

c_dic = prepCabinData(df)
cs_dic = prepCabinSurvData(df)
surv = []
nsurv = []
p_srv = []
for c in ages:
    surv.append(cs_dic.get(c, 0))
    nsurv.append(c_dic.get(c, 0))
    p_srv.append ((cs_dic.get(c,0)/((cs_dic.get(c,0) + c_dic.get(c,0)))*100))
    
data = {'Cabins' : ages,
        'Survived'   : surv,
        'Total'   : nsurv,
        'Percentage' : p_srv
       }

p = figure(x_range=ages, plot_height=250, title="Passenger Density & Survival w.r.t Cabins",
           toolbar_location=None, tools="", )

p.vbar_stack(years, x='Cabins', width=0.9, color=colors, source=data,
             legend=[value(x) for x in years])

p.y_range.start = 0
p.x_range.range_padding = 0.1
p.xgrid.grid_line_color = None
p.axis.minor_tick_line_color = None
p.outline_line_color = None
p.legend.location = "top_right"
p.legend.orientation = "horizontal"
p.xaxis.axis_label = "Cabins"
p.yaxis.axis_label = 'No. of Passengers'

hover = HoverTool(tooltips=[('Survival %age', '@Percentage')
                                    ],
                          mode='vline')
p.add_tools(hover)

show(p)

Loading BokehJS ...

The maximum survival percentage has been for the cabins E, B, D in the decreasing order (i.e 42%, 41.9%, 41%).  The cabin C was quite populated so the survival percentage is just 37%. There was only one passenger in cabin T who didn't survive.

## Titanic Dataset Findings: 

### Findings:
In the dataset, we found that majority of passengers were travelling to destinations in North America. When analyzing the survival rate, overall, the data showed a higher survival rate for first class passengers. Given that approximately 2/3 of the passengers died, third class passengers had a devestating mortality rate with 65%, where first class passengers fared much better with 15% mortality rate. Age and sex were fundamentally important in assessing the survival rate. Irrespective of sex, younger (0-10) and elderly (70+) saw much higher survival rates, both over 50%. Sex made a large difference, as the bar chart shows. Presumably, female passengers were evacuated first and thus their survival rates were always over 60%. 

Also the survival rate was better for passengers having 1 or 2 family members on board (about 51% and 45% respectively). For the people having more than 5 family members on board, the survival rate is almost zero. Most of the infants, kids and teenagers were in the 3rd class. This also tells us that the 3rd class was quite populated. The percentage of young adults and middle aged people in the 1st class is almost the same. There were large number of young adults in the 2nd class. There were a small number of elderly, all of them were in the third class. The maximum survival percentage has been for the cabins E, B, D. The data doesn't include the cabin layout of the ship but one can infer that these cabins must have been near the exit doors and/or the escape rafts so the people there had early access. The cabin C was quite populated so the survival percentage is just 37%. Cabin T was an outlier where survival percentage was zero.

### Post Mortem Analysis:
I found it difficult to draw conculsions on the survival rates as even the Third Class had 40%. Also the numbers fluctuate with the population size in different categories e.g one can infer that the number of infants in the third class was the largest, but it can simply be the case becuase the number of people travelling in the third class was the largest. It was difficult to draw intuitive conclusions from the correlations between some variables as they aren't seemingly related as we know them. Some data values were also missing from the dataset (e.g some passenger's cabin details) which made it hard to have a high confidence in the inference drawn.

On this particular dataset, it would be interesting to expand the data mining, perhaps develop a descision tree that could predict the survival rate on the discussed factors [age, sex, class, etc]. It is clear they are not independent and it could be interesting to see to what degree they influence each other. A multivariate analysis of the fields would also reveal some hidden aspects of the data which would allow us to understand (or estimate) the dynamic behavior/layout of the passengers which lead to their survival or death.



